fab/fabfile.py

```diff
@@ -718,18 +725,84 @@ def _deploy_without_asking():
         # hard update of manifest.json since we're about to force restart
         # all services
         _execute_with_timing(update_manifest)
+        _execute_with_timing(clean_releases)
+    except PreindexNotFinished:
+        mail_admins(
+            " You can't deploy yet",
+            ("Preindexing is taking a while, so hold tight "
+             "and wait for an email saying it's done. "
+             "Thank you for using AWESOME DEPLOY.")
+        )
     except Exception:
         _execute_with_timing(mail_admins, "Deploy failed", "You had better check the logs.")
         # hopefully bring the server back to life
         _execute_with_timing(services_restart)
         raise
     else:
+        _execute_with_timing(update_current)
         _execute_with_timing(services_restart)
         url = _tag_commit()
         _execute_with_timing(record_successful_deploy, url)
 
 
 @task
+@roles(ROLES_ALL_SRC)
+def update_current(release=None):
+    sudo('ln -nfs {} {}'.format(release or env.code_root, env.current))
+
+
+@task
+@roles(ROLES_ALL_SRC)
+def create_code_dir():
+    sudo('mkdir -p {}'.format(env.code_root))
+
+
+@task
+@roles(ROLES_ALL_SRC)
+def clean_releases(keep=3):
+    releases = sudo('ls {}'.format(env.releases))
+    to_remove = []
+    valid_releases = 0
+    for index, release in enumerate(releases):
+        if files.contains(RELEASE_RECORD, release):
+            valid_releases += 1
+            if valid_releases > keep:
+                to_remove.append(release)
+        else:
+            # cleans all releases that were not successful deploys
+            to_remove.append(release)
+
+    if len(to_remove) == len(release):
+        print 'Aborting, about to remove every release'
+        exit()
+
+    if os.path.basename(env.code_root) in to_remove:
+        print 'Aborting, about to remove current release'
+        exit()
+
+    for release in to_remove:
+        sudo('rm -rf {}{}'.format(env.releases, release))
+
+
+@task
+@roles(ROLES_ALL_SRC)
+def rollback():
```

Comment:
rollback needs some more thinking through. in the current state you could potentially rollback to a bad release. example:
RELEASES.txt
```
A <-- current
B
C
```
rollback to B
```
A 
B <-- current
C
```
deploy a new release X
```
X <-- current
A 
B 
C
```
rollback will now land you on A instead of B
```
X 
A <-- current
B 
C
```